File: day5/day5.py
def b2d(s):
  dec = 0
  for chr in s:
    dec = (dec*2+1) if (chr=='B' or chr=='R') else dec*2
  return dec
#print("Test ",b2d("BFFFBBFRRR"),"=567")
#print("Test ",b2d("FFFBBBFRRR"),"=119")
#print("Test ",b2d("BBFFBBFRLL"),"=820")

found_seats=[]
#avail_seats=[i for i in range(1024)]
input = open('day5/input.txt','r')
for line in input:
  s=b2d(line.strip())
  found_seats.append(s)
#  avail_seats.remove(s)

a=min(found_seats); b=max(found_seats)
seat_sum=(a+b)*(b-a+1)/2
my_seat=int(seat_sum-sum(found_seats))

print(max(found_seats),my_seat)

input.close()

File: day5/day5.hs
import System.IO

b2d :: [Char] -> Int
b2d s = helper $ reverse s
    where 
        helper [] = 0
        helper (c:cs)
            | c `elem` "BR" = 1 + 2 * (helper cs)
            | otherwise     = 2 * (helper cs)

seqsum :: (Int, Int) -> Int
seqsum (a,b) = (a+b)*(b-a+1) `div` 2

main = do
    input <- readFile "day5/input.txt"
    let seats = map b2d $ lines $ input
    print [b2d("BFFFBBFRRR"),567]
    print [b2d("FFFBBBFRRR"),119]
    print [b2d("BBFFBBFRLL"),820]
    let range = (minimum seats, maximum seats)
    let mySeat = (seqsum range) - sum seats
    print (range, mySeat)
